### PR TITLE
ddl: sync schema version using watch, notify sessions on owner node by job id

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #53246

Problem Summary:

### What changed and how does it work?
- sync schema version using etcd watch
- notify sessions on owner by job id to avoid triggering a wrong session when we run DDLs using multiple connections.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

1 pd/3 tikv, 1 tidb. Run on owner node. 20-30% faster

create table
```
thread=1, 500 tables
1m3.675s(127ms per table) -> 47.732s(95ms per table)

Thread=4, 2000 tables
3m8.39s(94ms per table) -> 2m20.273s(70ms per table)

Thread=16, 2000 tables
2m49.304s(85ms per table) -> 2m42.364s(81ms per table)
```
create database
```
thread=1, 500 databases
1m0.177s(120ms per op) -> 47.345s(95ms per op)

Thread=4, 2000 databases
3m35.87s(108ms per op) -> 2m29.758s(75ms per op)
```

1 pd/3 tikv, 8 tidb. Run on owner node, create table. 42% faster
```
thread=1, 500 tables
1m6s(132ms per table) -> 37.999s(76ms per table)
```

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/ddl/syncer/syncer.go [hunk 1]
```diff
@@ -126,16 +128,98 @@ type SchemaSyncer interface {
 	// the latest schema version. (exclude the isolated TiDB)
 	// It returns until all servers' versions are equal to the latest version.
 	OwnerCheckAllVersions(ctx context.Context, jobID int64, latestVer int64) error
+	// SyncJobSchemaVerLoop syncs the schema versions on all TiDB nodes for DDL jobs.
+	SyncJobSchemaVerLoop(ctx context.Context)
```
Comment (lance6716) [inline pkg/ddl/syncer/syncer.go:132]: can we move it inside NewSchemaSyncer?

File: pkg/ddl/syncer/syncer.go [hunk 2]
```diff
@@ -126,16 +128,98 @@ type SchemaSyncer interface {
...
+	v.Lock()
+	defer v.Unlock()
+	return len(v.nodeVersions)
+}
+
+func (v *nodeVersions) matchOrSet(fn func(nodeVersions map[string]int64) bool) {
+	v.Lock()
+	defer v.Unlock()
+	if ok := fn(v.nodeVersions); !ok {
+		// onceMatchFn must be nil before.
```
Comment (lance6716) [inline pkg/ddl/syncer/syncer.go:183]: maybe move it to function comments, so when I hover the cursor above the function in IDE I can see the comments.

File: pkg/ddl/syncer/syncer.go [hunk 3]
```diff
@@ -368,14 +467,150 @@ func (s *schemaVersionSyncer) OwnerCheckAllVersions(ctx context.Context, jobID i
...
+	// clean up emptyAndNotUsed items later.
+	for _, item := range s.jobNodeVersions {
+		item.clearData()
+	}
+	s.mu.Unlock()
+	for _, oneKV := range resp.Kvs {
+		s.handleJobSchemaVerKV(oneKV, mvccpb.PUT)
+	}
+	s.mu.Lock()
+	// we might miss some DELETE events during retry, some items might be emptyAndNotUsed, remove them.
```
Comment (lance6716) [inline pkg/ddl/syncer/syncer.go:514]: every `resp` of line 498 is a snapshot, we can always remove all the old data and apply the `resp`, no need to have two removing

File: pkg/ddl/ddl.go [hunk 4]
```diff
@@ -1205,13 +1233,18 @@ func (d *ddl) DoDDLJob(ctx sessionctx.Context, job *model.Job) error {
 		recordLastDDLInfo(ctx, historyJob)
 	}()
 	i := 0
+	notifyCh, ok := d.getJobDoneCh(job.ID)
+	if !ok {
+		// shouldn't happen, just give it a dummy one
```
Comment (lance6716) [inline pkg/ddl/ddl.go:1238]: it's OK to select from `nil` channel https://golangbyexample.com/select-with-nil-channel-golang/

**Review by lance6716 — COMMENTED** [review]
rest lgtm except for MDL part, I need more time to learn MDL behaviour 😂

File: pkg/ddl/syncer/syncer.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/ddl/syncer/syncer.go:132]: it's part of DDL, start background routines in `ddl.Start` should be more clear.

**Comment by ywqzzy** [comment]
The test result is unreliable, only 1 tidb..

File: pkg/ddl/syncer/syncer.go (hunk 3 above)
Comment (D3Hunter) [reply pkg/ddl/syncer/syncer.go]: https://github.com/pingcap/tidb/pull/53217/commits/738f11ca262bc7ca5e7472b1e8a561cc6adc1bf4

**Review by lance6716 — APPROVED** [review]

**Pull request merged**